src/chain/store/index.rs

```diff
@@ -151,139 +62,186 @@ impl<DB: Blockstore> ChainIndex<DB> {
+        }
+        if to > from.epoch() {
+            return Err(Error::Other(
+                "Looking for tipset with height greater than start point".to_string(),
+            ));
         }
-        let rounded = self.round_down(from)?;
-
-        let mut cur = rounded.key().clone();
-
-        loop {
-            let entry = self.skip_cache.lock().get(&cur).cloned();
-            let lbe = if let Some(cached) = entry {
-                metrics::LRU_CACHE_HIT
-                    .with_label_values(&[metrics::values::SKIP])
-                    .inc();
-                cached
-            } else {
-                metrics::LRU_CACHE_MISS
-                    .with_label_values(&[metrics::values::SKIP])
-                    .inc();
-                self.fill_cache(std::mem::take(&mut cur))?
-            };
 
-            if to == 0 {
-                if let Some(genesis_tipset_keys) =
-                    checkpoint_tipsets::genesis_from_checkpoint_tipset(lbe.tipset.key())
-                {
-                    let tipset = self.load_tipset(&genesis_tipset_keys)?;
-                    info!(
-                        "Resolving genesis using checkpoint tipset at height: {}",
-                        lbe.tipset.epoch()
-                    );
-                    return Ok(tipset);
-                }
+        for (child, parent) in self.chain(from).tuple_windows() {
+            if to == child.epoch() {
+                return Ok(child);
             }
-
-            if lbe.tipset.epoch() == to || lbe.parent_height < to {
-                return Ok(lbe.tipset.clone());
-            } else if to > lbe.target_height {
-                return self.walk_back(lbe.tipset.clone(), to);
+            if to > parent.epoch() {
+                // We're at a point where child.epoch() > x > parent.epoch().
+                match resolve {
+                    ResolveNullTipset::TakeOlder => return Ok(parent),
+                    ResolveNullTipset::TakeYounger => return Ok(child),
+                }
             }
-
-            cur = lbe.target.clone();
         }
+        Err(Error::Other(
+            "Tipset with epoch={to} does not exist".to_string(),
+        ))
     }
 
-    /// Walks back from the tipset, ignoring the cached entries.
-    /// This should only be used when the cache is checked to be invalidated.
-    pub(in crate::chain) fn get_tipset_by_height_without_cache(
-        &self,
-        from: Arc<Tipset>,
-        to: ChainEpoch,
-    ) -> Result<Arc<Tipset>, Error> {
-        self.walk_back(from, to)
+    /// Iterate from the given tipset to genesis. Missing tipsets cut the chain short.
+    pub fn chain(&self, from: Arc<Tipset>) -> impl Iterator<Item = Arc<Tipset>> + '_ {
+        itertools::unfold(Some(from), move |tipset| {
+            tipset.take().map(|child| {
+                *tipset = self.load_tipset(child.parents()).ok();
+                child
+            })
+        })
```

Comment:
deja vu: https://github.com/ChainSafe/forest/blob/a6ea63fb2ca4caab2e9d57a50d67af4289d24286/src/blocks/tipset.rs#L261-L268

Are these the same? If so, I think we have a problem

On one hand, this is a fairly trivial function, but on the other, I don't want to live in a codebase where there are multiple subtly different ways of doing the same thing. I think this needs
- consolidation in your PR
- some thinking about how to address the wider issue - what are your thoughts? What documentation could have led you to find the duplicate?

Reply:
Same semantics, but different performance. `Tipset::chain` is semantically the same as `ChainIndex::chain`, and `Tipset::load` is semantically the same as `ChainIndex::load_tipset`. The only difference is caching. We need both versions.

I'm open to suggestions for documentation and/or name changes.